controller/backup_target_controller.go

```diff
@@ -0,0 +1,313 @@
+	}
+}
+
+func (btc *BackupTargetController) processNextWorkItem() bool {
+	key, quit := btc.queue.Get()
+	if quit {
+		return false
+	}
+	defer btc.queue.Done(key)
+	err := btc.syncHandler(key.(string))
+	btc.handleErr(err, key)
+	return true
+}
+
+func (btc *BackupTargetController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync %v", btc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != btc.namespace {
+		// Not ours, skip it
+		return nil
+	}
+	return btc.reconcile(btc.logger, name)
+}
+
+func (btc *BackupTargetController) handleErr(err error, key interface{}) {
+	if err == nil {
+		btc.queue.Forget(key)
+		return
+	}
+
+	if btc.queue.NumRequeues(key) < maxRetries {
+		btc.logger.WithError(err).Warnf("Error syncing Longhorn backup target %v", key)
+		btc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	btc.logger.WithError(err).Warnf("Dropping Longhorn backup target %v out of the queue", key)
+	btc.queue.Forget(key)
+}
+
+func (btc *BackupTargetController) reconcile(log logrus.FieldLogger, name string) (err error) {
+	backupTarget, err := btc.ds.GetBackupTarget(name)
+	if err != nil {
+		if !datastore.ErrorIsNotFound(err) {
+			return err
+		}
+		log.Warnf("Backup target %s be deleted", name)
+		return nil
+	}
+
+	// Check the responsible node
+	if backupTarget.Spec.ResponsibleNodeID != btc.controllerID {
+		return nil
+	}
+
+	log = log.WithFields(logrus.Fields{
+		"url":      backupTarget.Spec.BackupTargetURL,
+		"cred":     backupTarget.Spec.CredentialSecret,
+		"interval": backupTarget.Spec.PollInterval.Duration,
+	})
+	backupTargetAvailable := backupTarget.Status.Available
+	backupTargetLastSyncedAt := backupTarget.Status.LastSyncedAt
+
+	defer func(backupTargetAvailable *bool, backupTargetLastSyncedAt **metav1.Time) {
+		if backupTargetAvailable == nil || backupTargetLastSyncedAt == nil {
+			log.Info("BUG: backupTargetAvailable or backupTargetLastSyncedAt is nil")
+			return
+		}
+
+		if backupTarget.Spec.BackupTargetURL == "" {
+			// Clean up all BackupVolume CRs
+			log.Info("Backup target URL is empty, clean up all the BackupVolume CRs")
+
+			clusterBackupVolumes, err := btc.ds.ListBackupVolume()
+			if err != nil {
+				log.WithError(err).Error("Error listing backup volumes in the cluster, proceeding with pull into cluster")
+			}
+			for backupVolumeName := range clusterBackupVolumes {
+				if err = btc.ds.DeleteBackupVolume(backupVolumeName); err != nil {
+					log.WithError(err).Errorf("Error deleting backup volume %s from cluster", backupVolumeName)
+				}
+			}
+		}
+
+		backupTarget.Status.Available = *backupTargetAvailable
+		backupTarget.Status.LastSyncedAt = *backupTargetLastSyncedAt
+		if _, err = btc.ds.UpdateBackupTargetStatus(backupTarget); err != nil && !datastore.ErrorIsConflict(err) {
+			log.WithError(err).Error("Error updating backup target status")
+		}
```

Comment:
For this part, you can follow the pattern in other controllers:
  1. Set the status field
  2. Do deep equal then determine if calling status update API is necessary.

Besides, there is no need to set the field `backupTarget.Status.Available = *backupTargetAvailable` here. If this field becomes false, the reconcile loop will return immediately.

Reply:
> Besides, there is no need to set the field `backupTarget.Status.Available = *backupTargetAvailable` here. If this field becomes false, the reconcile loop will return immediately.

Not quite understand this one. I'd want to update the `status.available` once in the reconcile loop, it detects the remote backup target suddenly unavailable.

Reply:
I mean, you don't need this line at all:
```
backupTarget.Status.Available = *backupTargetAvailable
```

Reply:
But there is probably the backup target available status from true to false.

Reply:
In the below code, you will set this status field to false then return from the current reconcile loop (then jump into this defer function) once the backup target is unavailable, right? Or in other words, once the reconcile loop enters into this defer function, `backupTarget.Status.Available` is already up-to-date. 

Similar issue to `backupTarget.Status.lastSyncedAt`